### DISCUSSION TITLE
Plugins which depend on other plugins

### DISCUSSION THREAD

**Discussion by chrisbrickhouse** [jspsych/jsPsych#1102] [Q&A]
I'm working on #1047 and #135, and I'm struggling to implement a modular structure where a wrapper plugin calls other plugins as submodules. As an example, say that we have `html-form-checkbox` which creates checkbox elements (and so on for other HTML input types), and we have a wrapper plugin `html-form`. I want to set up `html-form` so that it can use the `html-form-checkbox` interface to implement checkboxes without having to duplicate code across both. My questions are

1. How can I use a plugin inside a plugin?
1. Is this situation documented anywhere?
1. Do we have examples of plugins which depend on other plugins?

I didn't really know where to post this, so if discussions is the wrong place, let me know of a better venue.

**Comment by becky-gilbert** [comment]
Hi @chrisbrickhouse, great question. I think discussions is a good place for this, though issues would've worked too.

I'm not aware of any plugins that use other plugins. I think they were intentionally written to be private and stand-alone, with the exception that they of course depend on the core jspsych.js file. The core library does contain the [pluginAPI](https://www.jspsych.org/core_library/jspsych-pluginAPI/) helper functions that can used across plugins to avoid code duplication. But that doesn't seem like the right solution for what you're trying to do.

I'm not sure if what you're trying to do would be possible - hopefully @jodeleeuw or someone else can jump in!

**Comment by jodeleeuw** [comment]
I think the way to handle this is to define components for plugins as part of the core library, probably in a new module, and then use those components to construct plugins. 

We've talked about doing this for audio #341. I think there's also a discussion about implementing it for other kinds of components, but I can't recall where that is at the moment.
